docs/rfc/rfc-12.adoc

```diff
@@ -0,0 +1,430 @@
+the same coordination seed for the given coordination window.
+
+If heartbeat proposal is also not the case, the leader should send
+a **confirmation of activity**.
+
+To summarize, the pseudocode of the coordination leader routine is as follows:
+```
+if current_block % coordination_frequency == 0
+    propose_redemption_or_continue()
+
+    if current_block % (coordination_frequency * N) == 0
+        propose_deposit_sweep_or_continue()
+        propose_moved_funds_sweep_or_continue()
+
+    if current_block % (coordination_frequency * M) == 0
+        propose_moving_funds_or_continue()
+
+    rng = new RNG(coordination_seed)
+    if rng.get_bool_with_probability(12.5)
+        propose_heartbeat()
+
+    conirm_activity()
+```
+
+The coordination leader's routine can be implemented in the wallet client
+using the existing Go code living in the `pkg/maintainer/wallet` package
+of the `keep-core` repository.
+
+===== Communication
+
+Once the coordination leader completes the routine, it uses a dedicated
+secure broadcast channel to communicate the decision to the followers.
+It creates a **coordination message** and signs it using the operator's
+private key. A coordination message should look as follows:
+```
+CoordinationMessage
+    int member_id                // first signer controlled by the coordination leader
+    int coordination_block       // start block for the current coordination window
+    bytes wallet_public_key_hash // 20-byte public key hash of the wallet
+    bytes proposal               // serialized proposal; empty if this is just activity confirmation
+```
+
+Each signer-follower should listen for an incoming coordination message
+from the start of the coordination window until the coordination window ends.
+
+If a coordination message lands during the coordination window, the receiver
+should:
+
+- Make sure the given coordination message actually comes from the current
+  coordination leader. This can be done by knowing the coordination block,
+  coordinated wallet, and the coordination message's signature
+- Schedule the proposed wallet action to be executed at the coordination
+  window's end block. All necessary wallet state and proposal validation should
+  be done at the beginning of this action as a prerequisite of the threshold
+  signing process
+
+If the coordination message does not come from the current coordination leader,
+it should be ignored and its sender should be stored in the **coordination
+faults** cache.
+
+If the coordination message comes from the current coordination leader but
+carries an invalid proposal, it should be ignored and the current coordination
+leader should be stored in the **coordination faults** cache.
+
+If the coordination message does not come at all, the coordination window
+ends without any action. The current coordination leader who missed its turn
+should be stored in the **coordination faults** cache.
+
```

Comment:
> It looks like the message being signed [is already included in the session ID](https://github.com/keep-network/keep-core/blob/324f66fb3f1003f6cfeb7d4149ae3f1d902dba2e/pkg/tbtc/signing_loop.go#L251) of the announce protocol. This allows the signing group to exit the signing procedure early in case there is no consensus about the message being signed. 👌🏻

Yeah, session ID construction will nuke signing in this case. There is no possibility to solve this problem during the coordination message round (happens before signing and proposal validation) without using a consensus algorithm. Here we are just relying on the majority doing the same thing. If at least 51 members proceed with the received coordination message and execute signing successfully, the resulting state is considered as canonical.

> Also, the announcer is constructed [with a protocol name in the protocol ID](https://github.com/keep-network/keep-core/blob/324f66fb3f1003f6cfeb7d4149ae3f1d902dba2e/pkg/tbtc/signing.go#L216) which should address a potential problem of malicious leader sending one proposal to one set of followers and another proposal to another set of followers.

Not sure if I get this one. If a malicious leader does this, we are going back to the above case with session ID. Either the majority proceeds or everyone fail. What's the role of the protocol name in that case?

Reply:
> Not sure if I get this one. If a malicious leader does this, we are going back to the above case with session ID. Either the majority proceeds or everyone fail. What's the role of the protocol name in that case?

I am thinking of a situation when a malicious leader asks part of the network to sign a sweep transaction and part of the network to sign a moving funds transaction. So yes, this will fail as you described because the message to be signed will be different but *also* because those are two different actions and the protocol (action) ID is [a part of the announcement message as well](https://github.com/keep-network/keep-core/blob/324f66fb3f1003f6cfeb7d4149ae3f1d902dba2e/pkg/protocol/announcer/announcer.go#L119-L120.)

Reply:
> but also because those are two different actions and the protocol (action) ID is [a part of the announcement message as well](https://github.com/keep-network/keep-core/blob/324f66fb3f1003f6cfeb7d4149ae3f1d902dba2e/pkg/protocol/announcer/announcer.go#L119-L120.)

Negative. Protocols are `tbtc-dkg` or `tbtc-signing`. This is one level above wallet actions.

Reply:
Ah, right! In that case, we are fine anyway because the message being signed is included. Also, I agree with your comment [here](https://github.com/keep-network/tbtc-v2/pull/744#discussion_r1392535113):

>  I also think the coordination layer implementation must be aware of the wallet state to not naively schedule wallet actions that will be certainly nuked at the action validation stage.

but it's an implementation detail.